crates/sui-graphql-rpc/src/types/transaction_block.rs

```diff
@@ -297,102 +305,212 @@ impl TransactionBlock {
-                                .filter(tx_changed_objects::dsl::object_id.eq(o.into_vec()));
-                            query = query.filter(tx::dsl::tx_sequence_number.eq_any(sub_query));
-                        }
+        use checkpoints::dsl as cp;
+        use transactions::dsl as tx;
 
-                        if let Some(txs) = &filter.transaction_ids {
-                            let digests: Vec<_> = txs.iter().map(|d| d.to_vec()).collect();
-                            query = query.filter(tx::dsl::transaction_digest.eq_any(digests));
+        let (prev, next, transactions): (bool, bool, Vec<StoredTransaction>) = db
+            .execute_repeatable(move |conn| {
+
+                // The min or first `tx_sequence_number` of a checkpoint is the previous
+                // checkpoint's `network_total_transactions`. Because this refers to a historical
+                // checkpoint, if we yield a `None` result, we can return early.
+                let lo_tx = match lo_cp {
+                    0 => 0,
+                    _ => {
+                        let sequence_number: Option<i64> = conn
+                        .first(move || {
+                            cp::checkpoints
+                                .select(cp::network_total_transactions)
+                                .filter(cp::sequence_number.eq((lo_cp - 1) as i64))
+                        }).optional()?;
+
+                        match sequence_number {
+                            Some(sequence_number) => sequence_number as u64,
+                            None => return Ok::<_, diesel::result::Error>((false, false, vec![])),
                         }
-
-                        query
-                    },
-                )
+                    }
+                };
+
+                // The max or last `tx_sequence_number` of a checkpoint is the current checkpoint's
+                // `network_total_transactions` - 1. Since we cap the upper bound to
+                // `checkpoint_viewed_at`, we should always yield a result; thus, we can also return
+                // early if result is `None`.
+                let hi_tx = {
+                    let sequence_number: Option<i64> = conn
+                        .first(move || {
+                            cp::checkpoints
+                                .select(cp::network_total_transactions)
+                                .filter(cp::sequence_number.eq(hi_cp as i64))
+                        }).optional()?;
+
+                    match sequence_number {
+                        Some(sequence_number) => (sequence_number as u64).saturating_sub(1),
+                        None => return Ok::<_, diesel::result::Error>((false, false, vec![])),
+                    }
+                };
+
+                let sender = filter.sign_address;
+
+                let mut bound = TxLookupBound::from_range((Some(lo_tx), Some(hi_tx)));
```

Comment:
Flagging, because this is suspicious (but maybe an artifact of this being WIP) -- there should be scenarios where we don't have an upper or lower bound.